Tezos: support the new protocol update (005-BABYLON)

adonagy opened the pull request:
Hi!

This PR updates the firmware to be compatible with the new tezos protocol (https://tezos.gitlab.io/mainnet/protocols/005_babylon.html). The firmware is kept backward compatible with the current protocol to ensure proper functionality in case the protocol switch does not occur (not enough votes).

With these changes to the protocol, the binary representation of the operations have changed.

A major change is the deprecation of the scriptless kt1 accounts mainly used for token delegation. Now users can delegate tokens from the implicit (tz1, tz2 and tz3) accounts. After the protocol is approved and switched to, all the scriptless kt accounts will contain a master.tz script. This script will allow manipulation with the before scriptless accounts (set the delegate or transfer the tokens to an implicit account), thus creating new kinds of transaction operations. These transactions have an entrypoint with MICHELSON code prepended. 

Tested on the tezos-zeronet running the new protocol. Tests are included in the corresponding test file.

prusnak commented:
When is the launch of this hardfork expected?

adonagy commented:
In approximately 17 days, if voting participation reaches the quorum and the proposal has a super-majority in favor. 

https://tzscan.io/proposals

prusnak commented:
That's too bad. The first release we can fit this in is 6th of November ...

adonagy commented:
Unfortunately, they are not. In the proposed update the tags of the operations (transaction, reveal, origination, delegations) are changed to, quote: "make the incompatibility explicit between the two protocols"

matejcik commented:
is there a reason to support both protocols then?

prusnak commented:
> Is there a reason to support both protocols then?

Good point. I think it makes sense to rework the PR so only the post-hardfork functionality is available.

adonagy commented:
@matejcik @prusnak  Good point indeed.

The reason behind supporting both protocols was the synchronization between releases. In the case, that now is not possible, the firmware release happens before the protocol update

I'll rework the PR to support only the new protoco.

prusnak commented on common/protob/messages-tezos.proto:
```diff
@@ -60,7 +61,7 @@ message TezosSignTx {
      * Tezos contract ID
      */
-    message TezosContractID {
+     message TezosContractID {
```
We can remove `TezosContractID` completely if it's not used anymore at all.

adonagy replied:
We still use `TezosContractID` in the `TezosTransactionOp` as the type of the destination field

prusnak replied:
What's the difference between Implicit and Originated contract type?

adonagy replied:
Originated accounts(KT1) are created by the origination operation from an implicit account. The originated account can contain MICHELSON code, they're essentially smart contracts. 

Implicit accounts(tz{1, 2, 3}) are the 'classic' accounts, that are holding tokens and they can not include code (contract).

The contract type here is referring to the account type.

prusnak commented on common/protob/messages-tezos.proto:
```diff
@@ -75,7 +76,7 @@ message TezosSignTx {
      */
     message TezosRevealOp {
-        optional TezosContractID source = 1;
+        optional bytes source = 1;
```
Please change also the field number if you change the type.

`optional bytes source = 7;`

Same for other changes below.

adonagy replied:
Will do.

prusnak commented:
OK, so it seems the network has switched to Babylon.

Can you please test whether the firmware works with the new hardfork?

matejcik commented on common/protob/messages-tezos.proto:
```diff
@@ -60,7 +60,7 @@ message TezosSignTx {
      * Tezos contract ID
      */
-    message TezosContractID {
+     message TezosContractID {
```
this indent should be fixed

matejcik commented on core/src/trezor/messages/TezosTransactionOp.py:
```diff
@@ -3,6 +3,12 @@
 import protobuf as p
 
 from .TezosContractID import TezosContractID
+from .TezosTransactionSmartContractDelegationOp import (
```
not sure what happened here but this shouldn't be produced by `make gen`

adonagy commented:
> Can you please test whether the firmware works with the new hardfork?
 I'll test the firmware on the the updated Babylon protocol on the mainnet.  

In the meantime I've improved the the added operations, made it more readable and better to understand. I will update the PR ASAP, most likely tomorrow. I'll include the updated tests as well.

prusnak commented:
Great! I want to merge this code for our QA to test, so it would be best if we had the final changes on Wednesday the latest.

matejcik commented on core/src/apps/tezos/sign_tx.py:
```diff
@@ -215,3 +253,77 @@ def _encode_ballot(w: bytearray, ballot):
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["CONS"]))
+
+
+def _encode_smart_contract_delegation_remove(w: bytearray, smart_contract_delegation):
```
this function does not use the `smart_contract_delegation` argument. it should be removed

matejcik commented on core/src/apps/tezos/sign_tx.py:
```diff
@@ -215,3 +253,77 @@ def _encode_ballot(w: bytearray, ballot):
+
+    argument_length = (
+        sequence_length + 5
+    )  # 5 = tag and sequence_length (1 byte + 4 bytes)
```
please move the comment on the preceding line so that this doesn't happen

adonagy commented:
I've updated the PR with the refactor, added the last operation and tests. The checks are failing because the generated files seem little off. I'm allways using `make gen` for file regeneration. Don't know what's causing the problem. 

> I didn't understand the point of _encode_natural function, but I assume protocol-wise you know that it works ;)

It is almost identical with the zarith encoding. The difference being in setting the 7th bit if the value is negative. In the transfers however we never send negative amount of tokens. So the bit is never set but is still reserved for the sign.

matejcik commented on Pipfile:
```diff
@@ -64,4 +64,4 @@ attrs = "==18.2.0"  # let's use this version because of the nixos pipenv/pip iss
 
 [pipenv]
-allow_prereleases = true
+allow_prereleases = true
```
you're not leaving newlines at end of files. fix your editor config ;)

adonagy replied:
Oh, right. Will do.

matejcik commented on core/src/apps/tezos/sign_tx.py:
```diff
@@ -21,14 +21,43 @@
+                await layout.require_confirm_set_delegate(ctx, msg.transaction.fee)
+
+            # operation to remove delegate from the smart contract with manager.tz
+            if parameters_manager.cancel_delegate is not None:
```
does this mean that you can set and cancel delegate and perform a transfer in a single transaction?

adonagy replied:
That was not intended. I'll fix it

matejcik commented:
> I've updated the PR with the refactor, added the last operation and tests. The checks are failing because the generated files seem little off. I'm allways using `make gen` for file regeneration. Don't know what's causing the problem.

ISTM you fixed conflicts manually after the rebase instead of simply running `make gen` again?
I can push a commit that fixes it, but presumably if you run `make gen` on your side with the same result.

adonagy commented:
> ISTM you fixed conflicts manually after the rebase instead of running make gen again?
I can push a commit that fixes it, but presumably you can run make gen on your side with the same result.

Hmm, I'm quite sure I did run make gen after rebase, I'll double check. 

On an another note, whenever I run make gen the BITCOIN ONLY checks disappear in the core/src/trezor/messages/Capability.py and core/src/trezor/messages/Capability.py files and I have to fix the files manually.

matejcik commented:
> On an another note, whenever I run make gen the BITCOIN ONLY checks disappear in the core/src/trezor/messages/Capability.py and core/src/trezor/messages/Capability.py files and I have to fix the files manually.

The `BITCOIN_ONLY` check is added by a bunch of seds in `tools/build_protobuf`. `make gen` runs that. No idea why it wouldn't run for you.

matejcik commented:
https://github.com/trezor/trezor-firmware/blob/master/tools/build_protobuf#L72-L84

adonagy commented:
> The BITCOIN_ONLY check is added by a bunch of seds in tools/build_protobuf. make gen runs that. No idea why it wouldn't run for you.

Oh yeah, the seds. Always got a bunch of sed errors while running make gen, I should have checked into it. Looks like the problem was the sed on macOS. I did sort it out now. Everything is generating fine now. Thanks!

adonagy commented:
I've made the requested changes and fixed the generated files. If there is anything else, let me know.

prusnak commented:
I squashed the changes into 384275ac732c8c08a824cef82ed4b6a76355b122 and added one more small refactor to 322417d9ed2a0bfb6efdc8784c23a55702a1d042. Both of the commits are in the master now.

@adonagy Can you please test that my changes did not break the functionality? The tests pass, but testing in the wild is best of course.
